proposals/NNNN-dxil-vectors.md

```diff
@@ -0,0 +1,183 @@
+
+Although matrices are represented as vectors in some contexts such as unlinked library shaders,
+ their final DXIL representation will continue to be as arrays of scalars.
+This is consistent with both their past and future intended representation.
+
+### Changes to DXIL Intrinsics
+
+A new form of rawBufferLoad allows loading of full vectors instead of four scalars.
+The status integer for tiled resource access is loaded just as before.
+The returned vector value and the status indicator are grouped into a new `ResRet` helper structure type
+ that the load intrinsic returns.
+
+```asm
+  declare %dx.types.ResRet.v[NUM][TY] @dx.op.rawBufferLoad.v[NUM][TY](
+      i32,                  ; opcode
+      %dx.types.Handle,     ; resource handle
+      i32,                  ; coordinate c0 (index)
+      i32,                  ; coordinate c1 (elementOffset)
+      i8,                   ; mask
+      i32,                  ; alignment
+  )
+```
+
+The return struct contains a single vector and a single integer representing mapped tile status.
+
+```asm
+  %dx.types.ResRet.v[NUM][TY] = type { vector<TYPE, NUM>, i32 }
+```
+
+Here and hereafter, `NUM` is the number of elements in the loaded vector, `TYPE` is the element type name,
+ and `TY` is the corresponding abbreviated type name (e.g. `i64`, `f32`).
+
+#### Elementwise intrinsics
+
+A selection of elementwise intrinsics are given additional native vector forms.
+Elementwise intrinsics are those that perform their calculations irrespective of the location of the element
+ in the vector or matrix arguments except insofar as that position corresponds to those of the other elements
+ that might be used in the individual element calculations.
+An elementwise intrinsic `foo` that takes scalar or vector arguments could theoretically implement its vector version using a simple loop and the scalar intrinsic variant.
+
+```c++
+vec<TYPE, NUM> foo(vec<TYPE, NUM> a, vec<TYPE, NUM> b) {
+  vec<TYPE, NUM> ret;
+  for (int i = 0; i < NUM; i++)
+    ret[i] = foo(a[i], b[i]);
+}
+```
+  
+For example, `fma` is an elementwise intrinsic because it multiplies or adds each element of its argument vectors,
+ but `cross` is not because it performs an operation on the vectors as units,
+ pulling elements from different locations as the operation requires.
+
+The elementwise intrinsics that have native vector variants represent the
+ unary, binary, and tertiary generic operations:
+
+```asm
+ <[NUM] x [TYPE]> @dx.op.unary.v[NUM][TY](i32 opcode, <[NUM] x [TYPE]> operand1)
+ <[NUM] x [TYPE]> @dx.op.binary.v[NUM][[TY]](i32 opcode, <[NUM] x [TYPE]> operand1, <[NUM] x [TYPE]> operand2)
+ <[NUM] x [TYPE]> @dx.op.tertiary.v[NUM][TY](i32 opcode, <[NUM] x [TYPE]> operand1, <[NUM] x [TYPE]> operand2, <[NUM] x [TYPE]> operand3)
+```
+
+The only opcodes allowed with vector variants are:
+
+* Unary
+  * Exp
+  * Htan
+  * Atan
+  * Log
+* Binary
+  * FMin
+  * FMax
+* Tertiary
+  * Fma
```

Comment:
Just to point out, DXIL has both `fma` and `fmad`. The valid overload set for `fma` is fp64, where the valid overload set for `fmad` is all floating point types. I assume that `fma` in this list is intended to refer to both, but it would be good to be explicit.

Reply:
Makes sense. I was surprised at the tiny scope of opcodes listed here, but I assumed that the primary use case is not for doubles, so listing only `fma` without `fmad` confused me.

Reply:
This list is getting removed as we've determined that scalarized versus native elementwise intrinsics are an implementation detail.

Reply:
As long as there's sufficient test content that all vectorized intrinsics are covered, I'm fine without an exhaustive list in the spec.

Reply:
The exhaustive list is now the "allowed elementwise list" that was above this removed section.